cocotb/types/array.py

```diff
@@ -0,0 +1,314 @@
+
+        Args:
+            value: Any iterable of values that are constructable into elements using :meth:`_construct_elements`
+
+        Returns:
+            A :class:`list` of elements.
+
+        Raises:
+            TypeError: When the type isn't supported.
+            ValueError: When the iterable's values prevent construction into a list of elements.
+        """
+        return list(value)
+
+    @staticmethod
+    def _construct_element(elem: Any) -> Any:
+        """
+        Constructs a single element of an array.
+
+        For overriding by subclasses.
+        Used by __setitem__ when setting a single element.
+
+        Args:
+            elem: Any object that the implementation can turn into an array element.
+
+        Returns:
+            An array element.
+
+        Raises:
+            TypeError: When the type isn't supported.
+            ValueError: When the value prevents construction into an element.
+        """
+        return elem
+
+    @staticmethod
+    def _construct_range(rng: Any) -> Range:
+        """
+        Constructs the range portion of the array.
+
+        For overriding by subclasses.
+        Used by the constructor when setting the range
+        and by __setitem__ when creating a slice.
+
+        Args:
+            rng: Any object that the implementation can turn into a :class:`Range` object.
+
+        Returns:
+            The :class:`Range` object describing the array's indexing scheme.
+
+        Raises:
+            TypeError: When type isn't supported.
+            ValueError: When the value prevents construction into a range.
+        """
+        if isinstance(rng, Range):
+            return rng
+        raise TypeError("range argument must be of type 'Range'")
+
+    @cached_property
+    def left(self) -> int:
+        """Leftmost index of the array."""
+        return self._range.left
+
+    @cached_property
+    def direction(self) -> str:
+        """``'to'`` if indexes are ascending, ``'downto'`` otherwise."""
+        return self._range.direction
+
+    @cached_property
+    def right(self) -> int:
+        """Rightmost index of the array."""
+        return self._range.right
+
+    @cached_property
+    def length(self):
+        """Length of the array."""
+        return self._range.length
+
+    @cached_property
+    def range(self) -> Range:
+        """:class:`Range` of the indexes of the array."""
+        return self._range
+
+    def __len__(self) -> int:
+        return self.length
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._value)
+
+    def __reversed__(self) -> Iterator[Any]:
+        return reversed(self._value)
+
+    def __contains__(self, item: Any) -> bool:
+        return item in self._value
+
+    def __eq__(self, other: Any) -> bool:
+        try:
+            return all(a == b for a, b in zip_longest(self, other, fillvalue=object()))
+        except TypeError:
+            return NotImplemented
```

Comment:
This feels very python2-esque - silencing the TypeError seems like a bad idea to me.

Reply:
I think this is not right anyways.